src/Domain/FunctionsOfTime/ReadSpecThirdOrderPiecewisePolynomial.cpp

```diff
@@ -0,0 +1,98 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Domain/FunctionsOfTime/ReadSpecThirdOrderPiecewisePolynomial.hpp"
+
+#include <array>
+#include <cstddef>
+#include <map>
+#include <memory>
+#include <unordered_map>
+
+#include "DataStructures/Matrix.hpp"
+#include "Domain/FunctionsOfTime/PiecewisePolynomial.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/Dat.hpp"
+#include "IO/H5/File.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace importers {
+void read_spec_third_order_piecewise_polynomial(
+    gsl::not_null<std::unordered_map<
+        std::string, domain::FunctionsOfTime::PiecewisePolynomial<3>>*>
+        spec_functions_of_time,
+    const std::string& file_name,
+    const std::map<std::string, std::string>& dataset_name_map) noexcept {
+  // Currently, only support order 3 piecewise polynomials.
+  // This could be generalized later, but the SpEC functions of time
+  // that we will read in with this action will always be 3rd-order
+  // piecewise polynomials
+  constexpr size_t max_deriv{3};
+
+  std::unique_ptr<h5::H5File<h5::AccessType::ReadOnly>> file =
+      std::make_unique<h5::H5File<h5::AccessType::ReadOnly>>(file_name);
+  for (const auto& spec_and_spectre_names : dataset_name_map) {
+    const std::string& spec_name = std::get<0>(spec_and_spectre_names);
+    const std::string& spectre_name = std::get<1>(spec_and_spectre_names);
+    // clang-tidy: use auto when initializing with a template cast to avoid
+    // duplicating the type name
+    const h5::Dat& dat_file = file->get<h5::Dat>("/" + spec_name);  // NOLINT
```

Comment:
I thought on the line below I should use the type explicitly, since the typename Matrix isn't explicitly given on the RHS. Maybe I'm wrong here, but I thought we should avoid using auto& just to get out of specifying/thinking about what the type of a variable is? Is there more to your reasoning for wanting this to be `const auto& dat_data` instead of `const Matrix& dat_data`?

For the dat_file line, get<h5::Dat> clarifies the type, so I'm fine changing that one to auto

Reply:
That's not a guideline that I'm aware of. Our guidelines say: https://spectre-code.org/code_review_guide.html

> Almost always auto, except with expression templates, i.e. DataVector

I don't quite understand the part on expression templates, but we do currently use "almost always auto" in the codebase. The way I see it is that the compiler (and thus also the IDE and the developer) know the return type of the function from its declaration, so there's no need to specify it again at the call site, it's redundant.

Reply:
> except with expression templates, i.e. DataVector

This could have been more precisely written as "except when you want to cause an implicit conversion".  The most common case where you do is DataVector (or Blaze types in general).

Reply:
OK, I changed both this line and the next to `const auto&` in my next fixup commit